src/Compilers/CSharp/Portable/Symbols/Source/SourceModuleSymbol.cs

```diff
@@ -587,9 +613,27 @@ internal override bool NonNullTypes
         {
             get
             {
-                // PROTOTYPE(NullableReferenceTypes): temporary solution to avoid cycle
-                return SyntaxBasedNonNullTypes(((SourceAssemblySymbol)this.ContainingAssembly).GetAttributeDeclarations()) ?? this.UtilizesNullableReferenceTypes;
+                if (_lazyEarlyAttributeData == null)
+                {
+                    Interlocked.CompareExchange(ref _lazyEarlyAttributeData, ComputeEarlyAttributes(), null);
+                }
+                // PROTOTYPE(NullableReferenceTypes): REVIEW when no early attributes are there, _lazyEarlyAttributeData remains null and we re-do a bit of un-necessary work on every call
```

Comment:
>PROTOTYPE [](start = 19, length = 9)

@cston I wanted to check what you think here. I plan to remove this PROTOTYPE comment unless you also think this is a concern. Thanks

Reply:
Won't this be property be accessed frequently? If so, `ComputeEarlyAttributes()` should probably return an instance always.

---
In reply to: [201127385](https://github.com/dotnet/roslyn/pull/28320#discussion_r201127385) [](ancestors = 201127385)

Reply:
It looks like the other cycle mitigation I added removes the need for doing this early attributes gymnastics. I'll switch back to normal attribute bag logic, which is simpler.

---
In reply to: [201400408](https://github.com/dotnet/roslyn/pull/28320#discussion_r201400408) [](ancestors = 201400408,201127385)

Reply:
It turns out I was wrong, and the `NonNullTypesAttribute` still needs to be handled as an early attribute (some tests unrelated to nullable feature hit them in CI). I've reverted a commit and adjusted the caching logic so that the computation only happens once).